xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-code-oldcore/src/main/java/org/xwiki/rendering/internal/macro/code/source/DocumentAttachmentCodeMacroSourceLoader.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.rendering.internal.macro.code.source;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+import javax.inject.Singleton;
+
+import org.apache.commons.io.IOUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.rendering.internal.parser.pygments.PygmentsUtils;
+import org.xwiki.rendering.macro.MacroExecutionException;
+import org.xwiki.rendering.macro.code.source.CodeMacroSource;
+import org.xwiki.rendering.macro.code.source.CodeMacroSourceReference;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * @version $Id$
+ * @since 15.0RC1
+ * @since 14.10.2
+ */
+@Component(hints = {"ATTACHMENT", "PAGE_ATTACHMENT"})
+@Singleton
+public class DocumentAttachmentCodeMacroSourceLoader implements EntityCodeMacroSourceLoader
+{
+    @Override
+    public CodeMacroSource load(XWikiDocument document, EntityReference entityReference,
+        CodeMacroSourceReference reference, XWikiContext xcontext) throws MacroExecutionException
+    {
+        XWikiAttachment attachment = document.getAttachment(entityReference.getName());
+
+        if (attachment == null) {
+            throw new MacroExecutionException("Unknown attachment [" + entityReference + "]");
+        }
+
+        String content;
+        try (InputStream stream = attachment.getContentInputStream(xcontext)) {
```

Comment:
No restriction at all based on type? I mean someone could call it for an image, or a zip or anything?

Reply:
Sure, and the result won't be great for sure, but it's not really possible to know if the attachment contains text. Since it does not cause any security problem, I don't see the point of trying to be clever and error-prone here.

Reply:
Ok for the type. What about size of attachments? If someone tries to display a very big attachments, it could slow down the entire server no?

Reply:
Yes, if you ask for the content of a 10GB attachment, you will end up with an OOM. I guess we could have a configurable threshold, but not sure what default to use.

Reply:
@manuelleduc have you set a default value for the new attachment restrictions? If so we should maybe use same value.

Reply:
The default is no mimetype restriction. But indeed using `AttachmentValidator` could be considered.

Reply:
> But indeed using AttachmentValidator could be considered.

Isn't a stored attachment valid by definition ?

Reply:
I introduce a configurable maximum size, default being 10MB.